pkg/route/ingress/ingress.go

```diff
@@ -555,16 +564,30 @@ func newRouteForIngress(
 				{APIVersion: "networking.k8s.io/v1beta1", Kind: "Ingress", Controller: &t, Name: ingress.Name, UID: ingress.UID},
 			},
 		},
-		Spec: routev1.RouteSpec{
-			Host: rule.Host,
+	}
+	if hostIsWildcard {
+		route.Spec = routev1.RouteSpec{
+			Host: host,
+			Path: path.Path,
+			To: routev1.RouteTargetReference{
+				Name: path.Backend.ServiceName,
+			},
+			Port:           port,
+			TLS:            tlsConfigForIngress(ingress, rule, tlsSecret),
+			WildcardPolicy: routev1.WildcardPolicySubdomain,
+		}
+	} else {
+		route.Spec = routev1.RouteSpec{
+			Host: host,
 			Path: path.Path,
 			To: routev1.RouteTargetReference{
 				Name: path.Backend.ServiceName,
 			},
 			Port: port,
 			TLS:  tlsConfigForIngress(ingress, rule, tlsSecret),
-		},
+		}
 	}
```

Comment:
This can be simplified as follows:
```go
	var wildcardPolicy routev1.WildcardPolicyType
	if hostIsWildcard {
		wildcardPolicy = routev1.WildcardPolicySubdomain
	}
	t := true
	return &routev1.Route{
		ObjectMeta: metav1.ObjectMeta{
			GenerateName: ingress.Name + "-",
			Namespace:    ingress.Namespace,
			Labels:       ingress.Labels,
			Annotations:  ingress.Annotations,
			OwnerReferences: []metav1.OwnerReference{
				{APIVersion: "networking.k8s.io/v1beta1", Kind: "Ingress", Controller: &t, Name: ingress.Name, UID: ingress.UID},
			},
		},
		Spec: routev1.RouteSpec{
			Host: host,
			Path: path.Path,
			To: routev1.RouteTargetReference{
				Name: path.Backend.ServiceName,
			},
			Port:           port,
			TLS:            tlsConfigForIngress(ingress, rule, tlsSecret),
			WildcardPolicy: wildcardPolicy,
		},
	}
```

Reply:
the test are failing after adding this

--- FAIL: TestController_sync (0.00s)
    --- PASS: TestController_sync/no_changes (0.00s)
    --- PASS: TestController_sync/sync_namespace_-_no_ingress (0.00s)
    --- PASS: TestController_sync/sync_namespace_-_two_ingress (0.00s)
    --- PASS: TestController_sync/ignores_incomplete_ingress_-_no_host (0.00s)
    --- PASS: TestController_sync/ignores_incomplete_ingress_-_no_service (0.00s)
    --- PASS: TestController_sync/ignores_incomplete_ingress_-_no_paths (0.00s)
    --- PASS: TestController_sync/ignores_ingress_with_third-party_ingressclass (0.00s)
    --- PASS: TestController_sync/ignores_ingress_with_unsupported_path_type (0.00s)
    --- PASS: TestController_sync/ignores_incomplete_ingress_-_service_does_not_exist (0.00s)
    --- FAIL: TestController_sync/create_route (0.00s)
    --- FAIL: TestController_sync/create_route_-_targetPort_string,_service_port_with_name (0.00s)
    --- FAIL: TestController_sync/create_route_-_default_ingresscontroller (0.00s)
    --- FAIL: TestController_sync/create_route_-_custom_ingresscontroller (0.00s)
    --- FAIL: TestController_sync/create_wildcard_route_-_targetPort_string,_service_port_with_name (0.00s)
    --- FAIL: TestController_sync/update_wildcard_route_ (0.00s)
    --- PASS: TestController_sync/create_route_-_blocked_by_expectation (0.00s)
    --- FAIL: TestController_sync/update_route (0.00s)
    --- PASS: TestController_sync/no-op (0.00s)
    --- PASS: TestController_sync/no-op_-_ignore_partially_owned_resource (0.00s)
    --- PASS: TestController_sync/no-op_-_ignore_route_created_for_an_ingress_with_a_third-party_class (0.00s)
    --- PASS: TestController_sync/update_ingress_with_missing_secret_ref (0.00s)
    --- FAIL: TestController_sync/update_ingress_to_not_reference_secret (0.00s)
    --- FAIL: TestController_sync/update_route_with_old_owner_reference (0.00s)
    --- FAIL: TestController_sync/update_route_-_tls_config_missing (0.00s)
    --- FAIL: TestController_sync/update_route_-_termination_policy_changed_to_passthrough (0.00s)
    --- FAIL: TestController_sync/update_route_-_termination_policy_changed_to_reencrypt (0.00s)
    --- FAIL: TestController_sync/update_route_-_termination_policy_changed_to_reencrypt_and_no_tls_secret (0.00s)
    --- PASS: TestController_sync/termination_policy_on_ingress_invalid,_nothing_happens (0.00s)
    --- FAIL: TestController_sync/termination_policy_on_ingress_invalid,_disables_tls (0.00s)
    --- FAIL: TestController_sync/Empty_tlsconfig_enables_edge_termination_without_explicit_cert (0.00s)
    --- FAIL: TestController_sync/update_route_-_secret_values_changed (0.00s)
    --- PASS: TestController_sync/no-op_-_has_TLS (0.00s)
    --- PASS: TestController_sync/no-op_-_has_secret_with_empty_keys (0.00s)
    --- PASS: TestController_sync/no-op_-_termination_policy_has_been_changed_by_the_user (0.00s)
    --- PASS: TestController_sync/update_route_-_router_admitted_route (0.00s)
    --- PASS: TestController_sync/update_route_-_second_router_admitted_route (0.00s)
    --- PASS: TestController_sync/no-op_-_ingress_status_already_updated (0.00s)
    --- PASS: TestController_sync/no-op_-_router_rejected_route (0.00s)
    --- PASS: TestController_sync/delete_route_when_referenced_secret_is_not_TLS (0.00s)
    --- PASS: TestController_sync/delete_route_when_referenced_secret_is_not_valid (0.00s)
    --- PASS: TestController_sync/ignore_route_when_parent_ingress_no_longer_exists_(gc_will_handle) (0.00s)
    --- FAIL: TestController_sync/update_route_-_termination_policy_changed_to_passthrough_and_timeout_set (0.00s)
FAIL
FAIL	github.com/openshift/openshift-controller-manager/pkg/route/ingress	0.022s
FAIL
[miheer@localhost openshift-controller-manager]$

Reply:
The issue is happening because you are adding WildcardPolicy: wildcardPolicy everytime even when hostIsWildcard is false.
This is failing other tests as well. Do we want to set WildcardPolicy: routev1.WildcardPolicyNone by default on all routes ?

Reply:
@miheer the following simplification passes tests for me: 
```go
        var wildcardPolicy routev1.WildcardPolicyType
        if hostIsWildcard {
                wildcardPolicy = routev1.WildcardPolicySubdomain
        }
        t := true
        return &routev1.Route{
                ObjectMeta: metav1.ObjectMeta{
                        GenerateName: ingress.Name + "-",
                        Namespace:    ingress.Namespace,
                        Labels:       ingress.Labels,
                        Annotations:  ingress.Annotations,
                        OwnerReferences: []metav1.OwnerReference{
                                {APIVersion: "networking.k8s.io/v1", Kind: "Ingress", Controller: &t, Name: ingress.Name, UID: ingress.UID},
                        },
                },
                Spec: routev1.RouteSpec{
                        Host: host,
                        Path: path.Path,
                        To: routev1.RouteTargetReference{
                                Name: path.Backend.Service.Name,
                        },
                        Port:           port,
                        TLS:            tlsConfigForIngress(ingress, rule, tlsSecret),
                        WildcardPolicy: wildcardPolicy,
                },
        }
```
Looks like @Miciah 's solution just had some small typos (namely `Name: path.Backend.ServiceName,` & `{APIVersion: "networking.k8s.io/v1beta1"`)?